### PR TITLE
Improve coverage tools

### PR DESCRIPTION
**Pull request by funbringer**
This patch optimizes coverage data collection which leads to measurable reduction of CI times (~ -5 mins). However, I was disappointed to see that CircleCI doesn't handle incremental workspace updates very well (a couple of MiBs of coverage data still results in full-blown 1.6 GiB re-uploads; what a misfortune). Maybe @cicdteam knows something.

[Old](https://app.circleci.com/pipelines/github/zenithdb/zenith/4222/workflows/9aca6ccf-cdcb-4025-bcf3-8b6603f2c2d4/jobs/38586) vs [New](https://app.circleci.com/pipelines/github/zenithdb/zenith/4225/workflows/7391a2fb-d311-4548-b7f6-f3ab4764ed74/jobs/38623) (note the last two steps).

### CODE REVIEW

File: scripts/coverage [hunk 1]
```diff
@@ -205,55 +200,107 @@ class LLVM:
...
+# TODO: implement
+@dataclass
+class ProfDir:
+    cwd: Path
+    llvm: LLVM
+
+    def __post_init__(self) -> None:
+        self.cwd.mkdir(parents=True, exist_ok=True)
+
+    @property
```
Comment (LizardWizzard) [inline scripts/coverage:212]: if this field accessed several times it might be be good to utilize cached_property

File: scripts/coverage [hunk 2]
```diff
@@ -205,55 +200,107 @@ class LLVM:
         self._cov(subcommand='show', *extras, **kwargs)
 
 
+# TODO: implement
```
Comment (LizardWizzard) [inline scripts/coverage:203]: Done I guess? :)

File: scripts/coverage [hunk 3]
```diff
@@ -205,55 +200,107 @@ class LLVM:
...
+        if files_mtime >= profdata_mtime:
+            self.llvm.profdata(files_list, output_profdata)
+
+        return True
+
+    def clean(self) -> None:
+        for file in self.cwd.iterdir():
+            os.remove(file)
+
+    def __truediv__(self, other):
```
Comment (LizardWizzard) [inline scripts/coverage:244]: 👍🏻

File: scripts/coverage [hunk 4]
```diff
@@ -266,9 +313,9 @@ class HtmlReport(Report):
 @dataclass
 class GithubPagesReport(HtmlReport):
     output_dir: Path
-    commit_url: str
+    commit_url: str = 'https://local/deadbeef'
```
Comment (LizardWizzard) [inline scripts/coverage:316]: Probably unintentional

File: .circleci/config.yml [hunk 5]
```diff
@@ -340,6 +348,13 @@ jobs:
       # The store_test_results step tells CircleCI where to find the junit.xml file.
       - store_test_results:
           path: /tmp/test_output
+      - run:
+          name: Merge coverage data
+          command: |
+            # This will speed up workspace uploads
+            if [[ $BUILD_TYPE == "debug" ]]; then
+              scripts/coverage "--profraw-prefix=$CIRCLE_JOB" --dir=/tmp/zenith/coverage merge
```
Comment (LizardWizzard) [inline .circleci/config.yml:356]: $CIRCLE_JOB contains something that needs to be escaped?

**Review by LizardWizzard — APPROVED** [review]
Nice! Looks good to me, left few minor nitpicks

File: scripts/coverage (hunk 4 above)
Comment (funbringer) [reply scripts/coverage:316]: No, in fact, I added this to provide a fallback for local rendering. This dummy url will make it to the report.

File: scripts/coverage (hunk 1 above)
Comment (funbringer) [reply scripts/coverage:212]: No, I don't think that's a good idea. Sure, it's somewhat costly, but if used carefully, it won't do any harm at all. Just think of the subtle bugs that could happen with such a change.

Comment (LizardWizzard) [reply scripts/coverage:211]: You mean that file list can change on the fly? I think if this is the case many things will get broken anyway. Another option to better describe the intention might be to use get_files or similar name instead. For me `property` always means  something cheap to recreate every time it is accessed. It is not that important though, I suppose there wouldn't be too much files anyway

**Pull request merged**